### PR TITLE
WIP: Notebooks

### PR DESCRIPTION
**Pull request by Barisimre**
Add the two initial tutorial notebooks with TODO for comments

### CODE REVIEW

File: tutorials/basic_usage.ipynb [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    "conda activate lab-gatr\n",
+    "``` \n",
+    "Next, install PyTorch and xFormers and other libraries\n",
+    "```\n",
+    "pip install torch==2.1.0 --index-url https://download.pytorch.org/whl/cu121\n",
+    "pip install xformers==0.0.22.post7 --index-url https://download.pytorch.org/whl/cu121\n",
+    "pip install torch_geometric==2.4.0\n",
+    "pip install torch_scatter torch_cluster -f https://data.pyg.org/whl/torch-2.1.0+cu121.html\n",
+    "```\n",
+    "Install LaB-GATr itself, which also install GATr\n",
```
Comment (sukjulian) [inline tutorials/basic_usage.ipynb:24]: install**s** GATr

File: tutorials/basic_usage.ipynb [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    "conda create --name lab-gatr python=3.10\n",
+    "conda activate lab-gatr\n",
+    "``` \n",
+    "Next, install PyTorch and xFormers and other libraries\n",
+    "```\n",
+    "pip install torch==2.1.0 --index-url https://download.pytorch.org/whl/cu121\n",
+    "pip install xformers==0.0.22.post7 --index-url https://download.pytorch.org/whl/cu121\n",
+    "pip install torch_geometric==2.4.0\n",
+    "pip install torch_scatter torch_cluster -f https://data.pyg.org/whl/torch-2.1.0+cu121.html\n",
+    "```\n",
```
Comment (sukjulian) [inline tutorials/basic_usage.ipynb:23]: Maybe add a `git clone` here for the challenge submission.

File: tutorials/basic_usage.ipynb [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    "Additionally, if you have made a new Anaconda environment, install Jupyter\n",
+    "```\n",
+    "pip install jupyter jupyterlab\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
```
Comment (sukjulian) [inline tutorials/basic_usage.ipynb]: I always like to "Kernel > Restart & Clear Output" before I commit, then the notebook is fresh.

File: tutorials/detailed_model_reference.ipynb [hunk 4]
```diff
@@ -0,0 +1,663 @@
...
+   "execution_count": 2
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cb6a1fa784bd6801",
+   "metadata": {},
+   "source": [
+    "Next, a point cloud pooling transform for the tokenisation (patching).\n",
+    "\n",
+    "Nested hierarchy of sub-sampled point clouds. Each coarse-scale point is mapped to a cluster of fine-scale points. Barycentric interpolation from the coarse to the fine scales. For correct batching, ```torch_geometric.data.Data.__inc__()``` has to be overridden.\n",
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: I recently realised that what we are doing is not barycentric interpolation, so best to just write "interpolation".

File: tutorials/detailed_model_reference.ipynb [hunk 5]
```diff
@@ -0,0 +1,663 @@
...
+    "\n",
+    "```TODO```"
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: The point cloud pooling transform used by LaB-GATr pre-computes information required for coarsening and refinement of the input mesh.

For coarsening, we construct a [Voronoi](https://en.wikipedia.org/wiki/Voronoi_diagram)-type diagram. We first sample a fraction of the mesh vertices $\mathcal{P}$, approximately equidistantly, via farthest point sampling. Then we create the Voronoi clusters by querying, for each original mesh vertex, the nearest point in the sub-sampled set $\mathcal{P}$ via $k$-nearest neighbours ($k = 1$).

For interpolation, we simply query for each original mesh vertex the closest $k$ points in the sub-sampled set $\mathcal{P}$. For surface meshes, we choose $k = 3$ which corresponds to a triangular reference simplex, while for volume meshes we choose a tetrahedron ($k = 4$).

File: tutorials/detailed_model_reference.ipynb [hunk 6]
```diff
@@ -0,0 +1,663 @@
...
+   "id": "15893f75d230b52e"
+  },
+  {
+   "cell_type": "markdown",
+   "id": "546e0c748430850a",
+   "metadata": {},
+   "source": [
+    "A geometric algebra interface to embed your data in $\\mathbf{G}(3, 0, 1)$.\n",
+    "\n",
+    "```TODO```"
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: Simply giving input and output channels is not enough when working with geometric algebra. Instead, we must define how our data is going to be represented by multivectors (and auxiliary scalars). For convenience, we package this in a `GeometricAlgebraInterface`. Here, we embed the mesh surface normal as planes at each vertex and pass an addtional scalar feature.

At the model output, we "translate" the multivectors back into Euclidean vectors by interpreting them as translation (pun intended). There are multiple modelling choices for how to perform this extraction (see [paper](https://arxiv.org/abs/2403.07536)).

File: tutorials/detailed_model_reference.ipynb [hunk 7]
```diff
@@ -0,0 +1,663 @@
...
+  {
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-07-12T12:16:44.050245Z",
+     "start_time": "2024-07-12T12:16:44.045916Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "def interp(\n",
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb:248]: Maybe swap pooling and interpolation.

File: tutorials/detailed_model_reference.ipynb [hunk 8]
```diff
@@ -0,0 +1,663 @@
...
+   "outputs": [],
+   "execution_count": 5
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c832f5754fbd5911",
+   "metadata": {},
+   "source": [
+    "### 2.1 Tokenizer\n",
+    "```TODO```"
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: The interpolation layer performs a convex combination of multivectors which is compatible with the projective geometric algebra.

File: tutorials/detailed_model_reference.ipynb [hunk 9]
```diff
@@ -0,0 +1,663 @@
...
+    "    return mlp(multivectors, scalars, reference_mv=reference_multivector)"
+   ],
+   "id": "692225270547a492",
+   "outputs": [],
+   "execution_count": 6
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: The pooling layer is implemented via PyG message passing. We just have to embed the relative position between points as multivector.

File: tutorials/detailed_model_reference.ipynb [hunk 10]
```diff
@@ -0,0 +1,663 @@
...
+    "        return multivectors, scalars"
+   ],
+   "id": "dc8b7295bcc0eebb",
+   "outputs": [],
+   "execution_count": 7
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: The complete tokenization module now just wraps pooling and interpolation. For reflection equivariance, we have to keep track of a reference multivector. Furthermore, we support extracting a global class token for classification tasks. This is a lot more efficient than taking the global average of vertex-wise outputs.

File: tutorials/detailed_model_reference.ipynb [hunk 11]
```diff
@@ -0,0 +1,663 @@
...
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "### 2.2 LaB-GATr",
+   "id": "d8a241a8fdd731e6"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "```TODO```",
```
Comment (sukjulian) [inline tutorials/detailed_model_reference.ipynb]: Now we can assemble the complete LaB-GATr model. It is comprised of tokenization and a GATr backend. For correct handling of PyG-style batching, we use block-diagonal attention masks. Another highlight is the class token support which we implement via a `class_token_forward_wrapper`.

**Review by sukjulian — APPROVED** [review]
I like it a lot! This might be very useful to some.

File: tutorials/GRaM_lab_gatr.ipynb [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "## 2. Building the Model",
+   "id": "7abe7e5851297f92"
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "A geometric algebra interface to embed your data in $\\mathbf{G}(3, 0, 1)$.",
```
Comment (sukjulian) [inline tutorials/GRaM_lab_gatr.ipynb]: Here we can add the explanation from the previous review.

**Review by sukjulian — APPROVED** [review]
